### ISSUE TITLE
prefer-rest-params: ignore uses of ...arguments

### ISSUE THREAD

**Issue by sethkinast** [eslint/eslint#7130]
I would like to suggest a change to the **prefer-rest-params** rule.

My use case is thus:

```js
// constructor of this subclass takes arguments a, b, c, d, e, f
constructor(a) {
  super(...arguments);
  doSomethingSpecialWith(a);
}
```

In cases where a function takes many arguments, and you only want to operate on a subset, **prefer-rest-params** forces patterns like these:

```js
constructor(a, ...rest) {
  super(a, ...rest);
  doSomethingSpecialWith(a);
}

constructor(...args) {
  const [a] = args;
  super(...args);
  doSomethingSpecialWith(a);
}
```

The first pattern is not ideal because you have to change code in two places now if you change what variables you're using. The second pattern is not ideal in my opinion because it's added what seems to me to be mental overhead.

---

I would like to suggest that **prefer-rest-params** should not fire if the only use of `arguments` in a function is to spread it.

The following patterns would not be errors:

```js
constructor(a) {
  super(...arguments);
  doSomethingSpecialWith(a);
}
```

The following patterns would still be errors:

```js
constructor(a) {
  super(...arguments);
  doSomethingSpecialWith(arguments[0]);
}
```

**Comment by eslintbot** [comment]
Hi @sethkinast, thanks for the issue. It looks like there's not enough information for us to know how to help you.

If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

Requesting a rule change? Please see [Proposing a Rule Change](http://eslint.org/docs/developer-guide/contributing/rule-changes) for instructions.

If it's something else, please just provide as much additional information as possible. Thanks!

**Comment by platinumazure** [comment]
Hi @sethkinast, thanks for the issue. I can see this being a potentially worthwhile enhancement.

Unfortunately we do see dozens of issues a day, and it really helps us when issues follow a template. You have most of what we need, but can you please do me a favor and edit your post so it follows the "Proposing a Rule Change" template? Thanks so much!

**Comment by sethkinast** [comment]
The linked "Rule Change Proposal Template" at http://eslint.org/docs/developer-guide/contributing/rule-changes is 404

**Comment by platinumazure** [comment]
Um. Cripes, we did a restructure of all those templates recently, sorry.

I do apologize, I meant [this template](https://github.com/eslint/eslint/blob/master/templates/rule-change-proposal.md).

**Comment by platinumazure** [comment]
@sethkinast Can't repro, still getting 404?

Are you using an HTTPS Everywhere extension that might be forcing https? Unfortunately the ESLint site doesn't support HTTPS.

**Comment by sethkinast** [comment]
No https. The bot links to https://github.com/eslint/eslint/blob/master/templates/rule-change-proposal and under the heading "Proposing a Rule Change", the text "rule change proposal template" links to https://github.com/eslint/eslint/blob/master/templates/rule-change-proposal which 404s for me (no .md extension)

**Comment by sethkinast** [comment]
Sorry, pasted the wrong link for what the bot links to.

**Comment by sethkinast** [comment]
Incidentally, I initially followed the link in the issue template, which goes to http://eslint.org/docs/developer-guide/contributing/changes and makes no mention of a template for rule changes, other than a little list which I did try to follow.

**Comment by platinumazure** [comment]
Ah, I think I get it. I think our website strips off .md extensions for hyperlinks (which is okay when the link is still in eslint.org), but it shouldn't do that for external links (to GitHub).

[This template](https://github.com/eslint/eslint/blob/master/templates/rule-change-proposal.md) is the one we'll need. Thanks!

**Comment by sethkinast** [comment]
See initial issue post for justification, explanation, etc.

**What rule do you want to change?** prefer-rest-params


**Does this change cause the rule to produce more or fewer warnings?** Fewer


**How will the change be implemented? (New option, new default behavior, etc.)?** New default behavior


**Please provide some example code that this change will affect:**

```js
// parent constructor of this subclass takes arguments a, b, c, d, e, f
// but I only need to use a in my constructor
constructor(a) {
  super(...arguments);
  doSomethingSpecialWith(a);
}
```

**What does the rule currently do for this code?**
Error: Use the rest parameters instead of 'arguments'

**What will the rule do after it's changed?**
No error when the ONLY use of `arguments` is to spread it.

**Comment by platinumazure** [comment]
Thanks! I'll champion this. (Assumption is that if the operand of a spread
operator is arguments, allow it, else follow the current behavior.)

**Comment by mysticatea** [comment]
Thank you for this issue.

I'd like to see the concrete proposal. 

- I'm not sure it should relax for `constructor(x) { super(...arguments, a, b) }`. I don't see differences between that and `constructor(x, ...rest) { super(x, ...rest, a, b) }`.
- I think `constructor(...args) { super(...arguments) }` this `arguments` should be warned.
- I think `constructor() { super(...arguments) }` this `arguments` should be warned.

So it sounds good to me that the rule ignores `arguments` under the following conditions:

- A function has parameters except for rest parameters.
- In the function, `...arguments` is the single argument of `CallExpression`.

**Comment by sethkinast** [comment]
Interesting, thank you for the detailed thoughts.

For 1, I need to think about it more.
For 2, I agree.
For 3, I am not sure I agree-- I think this is the explicit case where the developer does not care about the arguments at all except to pass them along. This one is one of the primary drivers of my current issue. Could you elaborate on your reason for this one?

**Comment by mysticatea** [comment]
> Could you elaborate on your reason for this one?

Just it's the purpose of the rule :smile: 
This rule prefers rest parameters than `arguments`.

**Comment by sethkinast** [comment]
Haha, fair enough. I expected that the purpose of the rule was to avoid confusing uses of arguments (hence `prefer-` instead of `no-arguments`), but if that's the case I will probably not try to take this forward because it would not solve my primary problem with the rule.

I don't want to allow direct arguments property lookups, but I don't see the need to alias `arguments` just to pass it somewhere else.

Thank you for the prompt feedback.

**Comment by mathieug** [comment]
With "prefer-rest-params: "error", how should I fix this case:

```javascript
module.exports = function coroute(gen) {
  const fn = bluebird.coroutine(gen);

  // Error handler signature: err, req, res, next
  // Param handler signature: req, res, next, param
  function handleErrorOrParam(err, req, res, next) {
    const nextFn = _.isError(err) ? next : res;

    return fn(...arguments).catch(nextFn); // → Use the rest parameters instead of 'arguments'
  }

  // Route signature: req, res, next
  function handleRequest(req, res, next) {
    return fn(...arguments).catch(next); // → Use the rest parameters instead of 'arguments'
  }

  return gen.length >= handleErrorOrParam.length ? handleErrorOrParam : handleRequest;
};
```

**Comment by vitorbal** [comment]
@mathieug please open a new issue or drop by our [Gitter chat](https://gitter.im/eslint/eslint) if you need any help. Thanks!
